pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,144 @@
+	"sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+
+	"istio.io/istio/pilot/pkg/model"
+	serviceRegistryKube "istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+)
+
+type ServiceExportController struct {
+	client        versioned.Interface
+	serviceClient corev1.CoreV1Interface
+
+	queue           queue.Instance
+	serviceInformer cache.SharedInformer
+
+	environment *model.Environment
+
+	// We use this flag to short-circuit the logic and stop the controller if the CRD does not exist (or is deleted)
+	crdExists bool
+}
+
+func NewServiceExportController(kubeClient kube.Client, environment *model.Environment) (*ServiceExportController, error) {
+	serviceExportController := &ServiceExportController{
+		client:        kubeClient.MCSApis(),
+		serviceClient: kubeClient.Kube().CoreV1(),
+		queue:         queue.NewQueue(time.Second),
+		environment:   environment,
+		crdExists:     true,
+	}
+
+	serviceExportController.serviceInformer = kubeClient.KubeInformer().Core().V1().Services().Informer()
+	serviceExportController.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { serviceExportController.onServiceAdd(obj) },
+		DeleteFunc: func(obj interface{}) { serviceExportController.onServiceDelete(obj) },
+	})
+
+	return serviceExportController, nil
+}
+
+func (sc *ServiceExportController) onServiceAdd(obj interface{}) {
+	sc.queue.Push(func() error {
+		serviceObj, err := convertToService(obj)
+		if err != nil {
+			return err
+		}
+		return sc.handleNewService(serviceObj)
+	})
+}
+
+func (sc *ServiceExportController) onServiceDelete(obj interface{}) {
+	sc.queue.Push(func() error {
+		serviceObj, err := convertToService(obj)
+		if err != nil {
+			return err
+		}
+		return sc.handleDeletedService(serviceObj)
+	})
+}
+
+func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
+	cache.WaitForCacheSync(stopCh, sc.serviceInformer.HasSynced)
+	log.Infof("Syncing existing services and serviceexports...")
+	sc.doInitialFullSync()
+	log.Infof("serviceexport sync complete")
+	log.Infof("ServiceExport controller started")
+	go sc.queue.Run(stopCh)
+}
+
+func (sc *ServiceExportController) handleNewService(svc *v1.Service) error {
+	if !sc.crdExists {
+		return nil
+	}
+	hostname := serviceRegistryKube.ServiceHostname(svc.Name, svc.Namespace, sc.environment.DomainSuffix)
+	if sc.environment.ClusterLocal().GetClusterLocalHosts().IsClusterLocal(hostname) {
+		return nil // don't do anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(svc)
+}
+
+func (sc *ServiceExportController) handleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func (sc *ServiceExportController) createServiceExportIfNotPresent(service *v1.Service) error {
+	serviceExport := v1alpha1.ServiceExport{}
+	serviceExport.Namespace = service.Namespace
+	serviceExport.Name = service.Name
```

Comment:
You should consider using an `OwnerReference` too.  You won't have to worr about deletion when the owning `Service` is deleted.  Also, if you're going to manage deletion, you might want to add a finalizer entry on the `Service` to ensure you don't miss the deletion event.